### ISSUE TITLE
define pairs for Cons

### ISSUE THREAD

**Issue by jlapeyre** [JuliaCollections/DataStructures.jl#580]
With this definition
```julia
Base.pairs(x::Cons) = (first(p) => last(p) for p in enumerate(x))
```
`findall`, `findfirst`, etc.  will work with linked lists.

Or maybe even, since there seems to be no default method for `pairs`,
```julia
Base.pairs(x) = (first(p) => last(p) for p in enumerate(x))
```

**Comment by jlapeyre** [comment]
And this allows `findlast` to work with `Cons`
```julia
Base.Iterators.Reverse(x::Cons) = reverse(x)
```

**Comment by oxinabox** [comment]
Oh thats pretty cunning.
Does `Base.pairs(x::Cons) = enumerate(x)`  work?

**Comment by jlapeyre** [comment]
> Does Base.pairs(x::Cons) = enumerate(x) work?

Yes. Good eye. I wonder then if  `Base.pairs(x) = enumerate(x)` is a good idea.  This would assume that any type that has not defined a method for `pairs`, and that can be iterated over, can sensibly be indexed by one-dimensional `LinearIndices`.  On the other hand, new types could come to rely on this without realizing it, which would make it hard to take back later if it turns out to be a bad idea. At a minimum, defining the method for `x::Cons` makes sense.

**Comment by jlapeyre** [comment]
I wonder if `Cons` is actually used by anyone, or if this is all just for fun.  This came up because I was thinking of  a demonstration of the advantages of generic functions in Julia compared to Common Lisp. Since singly-linked lists are so common in in CL, I wanted to show that `findall` compiles to efficient code for many data structures, including lists. But, it didn't work. Of course, demonstrating how easy it is to fix is impressive as well.

**Comment by oxinabox** [comment]
we have a few other packages for this kind of list.
FunctionalCollections.jl, and Lists.jl

It might be a good idea to remove this from DataStructures and leaver it to a package that is more into that kind of collection;
to thus decrease our code surface to get ready to release 1.0

**Comment by karan0299** [comment]
So, in the above discussion should I infer that `Cons` DataStructure from this package?
If so, I would like to work on this.
Otherwise also, I would like to take up this issue.
Thanks!

**Comment by oxinabox** [comment]
No decision has been made here. Yet.
Options:
1. Remove `Cons` from this package (potentially putting it elsewhere)
2. Add `pairs` definition to this package
3. Add fallback `pairs` definition to Base

**Comment by karan0299** [comment]
I think 2nd option would be best  in this scenario
For 3rd option I think, I have to change the Julia/JuliaLang repo.
For now, I cannot figure out where to put `Cons` if we remove from this package.

**Comment by jlapeyre** [comment]
It might be worth mentioning this in a Julia language issue.

**Comment by oxinabox** [comment]
@jlapeyre  feel encouraged to open such an issue and link to here
